Firestore/core/src/bundle/bundle_loader.cc

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/bundle/bundle_loader.h"
+
+#include <memory>
+#include <unordered_map>
+
+#include "Firestore/core/include/firebase/firestore/firestore_errors.h"
+#include "Firestore/core/src/api/load_bundle_task.h"
+#include "Firestore/core/src/bundle/bundle_document.h"
+#include "Firestore/core/src/model/document_key.h"
+#include "Firestore/core/src/model/document_key_set.h"
+#include "Firestore/core/src/model/model_fwd.h"
+#include "Firestore/core/src/model/no_document.h"
+
+namespace firebase {
+namespace firestore {
+namespace bundle {
+
+using firestore::Error;
+using firestore::api::LoadBundleTaskProgress;
+using firestore::api::LoadBundleTaskState;
+using model::DocumentKeySet;
+using model::MaybeDocumentMap;
+using model::NoDocument;
+using util::Status;
+using util::StatusOr;
+
+StatusOr<absl::optional<LoadBundleTaskProgress>> BundleLoader::AddElement(
+    std::unique_ptr<BundleElement> element_ptr, uint64_t byte_size) {
+  HARD_ASSERT(element_ptr->element_type() != BundleElement::Type::Metadata,
+              "Unexpected bundle metadata element.");
+
+  auto before_count = documents_.size();
+
+  const auto& element = *element_ptr;
+  if (element.element_type() == BundleElement::Type::NamedQuery) {
+    queries_.push_back(static_cast<const NamedQuery&>(element));
+  } else if (element.element_type() == BundleElement::Type::DocumentMetadata) {
+    const auto& document_metadata =
```

Comment:
This function is pretty long. I'd suggest moving the handling of metadata and of document to helper functions (the one for document should probably return a `Status`).

Reply:
These two branches sorta relate to each other, if metadata has exist()==true, it sets `current_document_`, otherwise it inserts the document directly, and in the branch for document this is checked agains.

I think it makes sense to keep them together in this case?

Reply:
Ok, how about moving this whole `if` statement into a separate function? That way, both functions will be of a more manageable size:
```cpp
  auto before_count = documents_.size();

  ProcessElement(*element_ptr); // Feel free to use a better name
  bytes_loaded_ += byte_size;

  if (before_count == documents_.size()) {
    return {absl::nullopt};
  }
  // ...
```

Additional suggestions:
* add empty lines before each `else if` clause -- this would help to visually separate the different branches;
* consider switching to `switch` (the slight advantages are less verbose conditions in each branch and the compiler potentially checking that the switch is exhaustive).